### ISSUE TITLE
Incorrect mapper code generated for mapping from a single property to property of collection element

### ISSUE THREAD

**Issue by adammichalik** [orika-mapper/orika#104]
Orika 1.4.6
Origin: http://stackoverflow.com/q/35084727/466738

Test case:

``` java
public class A {
    String stringA;

    public String getStringA() {
        return stringA;
    }

    public void setStringA(String stringA) {
        this.stringA = stringA;
    }
}

public class B {
    Set<BSetElement> elements = new HashSet<>();

    public Set<BSetElement> getElements() {
        return elements;
    }

    public void setElements(Set<BSetElement> elements) {
        this.elements = elements;
    }
}

public class BSetElement {
    String stringB;

    public String getStringB() {
        return stringB;
    }

    public void setStringB(String stringB) {
        this.stringB = stringB;
    }
}

// ---- test
// assertions from import static org.assertj.core.api.Assertions.*;

    @Test
    public void test() {
        MapperFactory mapperFactory = new DefaultMapperFactory.Builder().build();
        mapperFactory.classMap(A.class, B.class)
                .field("stringA", "elements{stringB}")
                .register();
        A a = new A();
        a.setStringA("test");

        B b = mapperFactory.getMapperFacade().map(a, B.class);

        assertThat(b.getElements()).extracting("stringB").containsExactly("test");
    }
```

Error:

```
15:28:57.213 [main] DEBUG m.g.orika.metadata.ClassMapBuilder - ClassMap created:
    ClassMapBuilder.map(A, B)
     .field( stringA(String), elements{stringB}(String) )
15:28:57.694 [main] DEBUG m.g.o.impl.generator.MapperGenerator - Generating new mapper for (A, B)
    Orika_B_A_Mapper21832028326451$0.mapAToB(A, B) {
     Field(stringA(String), elements{stringB}(String)) : copying String by reference
    }
    Orika_B_A_Mapper21832028326451$0.mapBToA(B, A) {
     Field(elements{stringB}(String), stringA(String)) : copying String by reference
    }
15:28:57.698 [main] DEBUG m.g.orika.impl.MapperFacadeImpl - MappingStrategy resolved and cached:
    Inputs:[ sourceClass: orika.A, sourceType: null, destinationType: class orika.B]
    Resolved:[ strategy: InstantiateAndUseCustomMapperStrategy, sourceType: A, destinationType: B, mapper: GeneratedMapper<A, B> {usedConverters: [], usedMappers: [], usedMapperFacades: [], usedTypes: [] }, mapReverse?: false]

ma.glasnost.orika.MappingException: While attempting the following mapping:
sourceClass = class orika.A
destinationType = orika.B
resolvedStrategy = InstantiateAndUseCustomMapperStrategy<A, B> {customMapper: GeneratedMapper<A, B> {usedConverters: [], usedMappers: [], usedMapperFacades: [], usedTypes: [] }, unenhancer: ma.glasnost.orika.unenhance.BaseUnenhancer@fa49800, objectFactory: DefaultConstructorObjectFactory<B>}
Error occurred: java.lang.NullPointerException

-----begin dump of current state-----------------------------
Registered object factories: 1 (approximate size: 24.6 kB)
  [B] : {A=DefaultConstructorObjectFactory<B>}
-------------------------------------------------------------
Registered mappers: 1 (approximate size: 4,108.0 kB)
  [0] : GeneratedMapper<A, B> {usedConverters: [], usedMappers: [], usedMapperFacades: [], usedTypes: [] }
-------------------------------------------------------------
Registered concrete types: 5 (approximate size: 108.8 kB)
  [interface java.util.Collection] : ArrayList<Object>
  [interface java.util.Map] : LinkedHashMap<Object, Object>
  [interface java.util.Set] : LinkedHashSet<Object>
  [interface java.util.List] : ArrayList<Object>
  [interface java.util.Map$Entry] : MapEntry<Object, Object>
-------------------------------------------------------------
Resolved strategies: 1 (approximate size: 4,625.6 kB)
{source: A, dest: B, in-place:false}: InstantiateAndUseCustomMapperStrategy<A, B> {customMapper: GeneratedMapper<A, B> {usedConverters: [], usedMappers: [], usedMapperFacades: [], usedTypes: [] }, unenhancer: ma.glasnost.orika.unenhance.BaseUnenhancer@fa49800, objectFactory: DefaultConstructorObjectFactory<B>}
-------------------------------------------------------------
Unenhance strategy: ma.glasnost.orika.unenhance.BaseUnenhancer@fa49800
-----end dump of current state-------------------------------
    at ma.glasnost.orika.impl.ExceptionUtility.newMappingException(ExceptionUtility.java:55)
    at ma.glasnost.orika.impl.MapperFacadeImpl.map(MapperFacadeImpl.java:752)
    at ma.glasnost.orika.impl.MapperFacadeImpl.map(MapperFacadeImpl.java:721)
    at orika.OrikaTest.test(OrikaTest.java:42)
[...]
Caused by: java.lang.NullPointerException
    at ma.glasnost.orika.generated.Orika_B_A_Mapper22107286298521$0.mapAtoB(Orika_B_A_Mapper22107286298521$0.java:20)
    at ma.glasnost.orika.impl.mapping.strategy.UseCustomMapperStrategy.map(UseCustomMapperStrategy.java:67)
    at ma.glasnost.orika.impl.MapperFacadeImpl.map(MapperFacadeImpl.java:742)
    ... 29 more
```

It seems the generated code is incorrect:

``` java
public class Orika_B_A_Mapper22107286298521$0 extends
        ma.glasnost.orika.impl.GeneratedMapperBase {

    public void mapAtoB(java.lang.Object a, java.lang.Object b,
            ma.glasnost.orika.MappingContext mappingContext) {

        super.mapAtoB(a, b, mappingContext);

        // sourceType: A
        orika.A source = ((orika.A) a);
        // destinationType: B
        orika.B destination = ((orika.B) b);

        java.util.Set new_destinationElements0 = null;
        new_destinationElements0 = null;
        orika.BSetElement elements_destination0Element = null;
        boolean elements_destination0ElementShouldBeAddedToCollector = false;
/* 20: */   elements_destination0Element.setStringB(((java.lang.String) source
                .getStringA()));
        if (!(new_destinationElements0 == null)
                && !new_destinationElements0.isEmpty()) {
            if (((java.util.Set) destination.getElements()) == null) {
                destination
                        .setElements(((java.util.Set) new java.util.LinkedHashSet()));
            } else {

                ((java.util.Set) destination.getElements()).clear();
            }

            ((java.util.Set) destination.getElements())
                    .addAll(new_destinationElements0);
        }

        if (customMapper != null) {
            customMapper.mapAtoB(source, destination, mappingContext);
        }
    }

    public void mapBtoA(java.lang.Object a, java.lang.Object b, ma.glasnost.orika.MappingContext mappingContext) { ... }
}
```

It's clear that `elements_destination0Element` is `null` at line 20 as declared and assigned 2 lines above.
The same error happens if the collection property is of type `List`.

**Comment by zafar-iqbal** [comment]
Can me know, When this issue will be fixed? 
single property to property of collection element.

**Comment by brabenetz** [comment]
I think the 'NullpointerException' is fixed by #175.
But the Example Code will only work for mapping to an existing instance which already have one or more elements (works after Bugfix #175 ):
```java
MapperFactory mapperFactory = new DefaultMapperFactory.Builder().build();
mapperFactory.classMap(A.class, B.class)
	.field("stringA", "elementsSet{stringB}")
	.register();

A a = new A();
a.setStringA("test");

// Instead of: B b = mapperFactory.getMapperFacade().map(a, B.class);
B b = new B();
b.getElementsSet().add(new BSetElement());
mapperFactory.getMapperFacade().map(a, b);

assertThat(b.getElementsSet(), hasSize(1));
assertThat(b.getElementsSet().iterator().next().getStringB(), is("test"));
```

A Workaround could be to use the indexed-Expression, but this is only Supported if the Collection is a java.util.List:
```java
MapperFactory mapperFactory = new DefaultMapperFactory.Builder().build();
mapperFactory.classMap(A.class, B.class)
	.field("stringA", "elementsList[0].stringB")
	.register();

A a = new A();
a.setStringA("test");

B b = mapperFactory.getMapperFacade().map(a, B.class);

assertThat(b.getElementsList(), hasSize(1));
assertThat(b.getElementsList().iterator().next().getStringB(), is("test"));
```

Another Workaround can be a CustomConverter:
```java
MapperFactory mapperFactory = new DefaultMapperFactory.Builder().build();
mapperFactory.classMap(A.class, B.class)
	.customize(new CustomMapper<A, B>() {
		@Override
		public void mapAtoB(A a, B b, MappingContext context) {
			if (StringUtils.isNotEmpty(a.getStringA())) {
				BSetElement bElem = new BSetElement();
				bElem.setStringB(a.getStringA());
				b.getElementsSet().add(bElem);
			}
		}
	})
	.register();
A a = new A();
a.setStringA("test");

B b = mapperFactory.getMapperFacade().map(a, B.class);

assertThat(b.getElementsSet(), hasSize(1));
assertThat(b.getElementsSet().iterator().next().getStringB(), is("test"));
```
Here is the TestCase I used: 
[Issue104TestCase.java.txt](https://github.com/orika-mapper/orika/files/680718/Issue104TestCase.java.txt)

**Comment by andreabertagnolli** [comment]
Agree with @brabenetz

**Comment by andreabertagnolli** [comment]
Fixed by #180
